File: scripts/superkey.py
#!/usr/bin/env python3
"""
MATCH-LEVEL SUPER-KEY — the honest event-clustering key (truth-audit attack E).

The incumbent cluster key is `COALESCE(event_slug, condition_id)`. But event_slug only
groups OUTCOME variants of ONE sub-market:

    slug fifwc-civ-nor-2026-06-30-exact-score-3-0  → event_slug fifwc-civ-nor-2026-06-30-exact-score

A single real-world match spans SEVERAL event_slugs — game-winner, exact-score, "more-markets"
(spreads/totals/corners/halves), draw, etc.:

    fifwc-bel-sen-2026-07-01               (game winner)
    fifwc-bel-sen-2026-07-01-exact-score   (6 exact-score outcomes)
    fifwc-bel-sen-2026-07-01-more-markets  (spreads/totals/corners)
    ''  (empty event_slug; slug = fifwc-bel-sen-2026-07-01-total-1pt5)

Clustering on event_slug therefore treats one match as up to ~4 "independent events" → inflates
N, tightens the confidence bound, and lifts z. The honest unit is the MATCH.

Rule (documented, deterministic, no fitting):
  base = event_slug if non-empty else slug            # empty event_slug rows carry the match in slug
  if base contains a YYYY-MM-DD date:
      super_key = base up-to-and-INCLUDING the first date   # everything after the date is market-type noise
  else:
      super_key = base                                 # dateless markets (elections, crypto up/down) stay 1:1

Why "up to the first date" is safe: every sports slug is `<sport>-<t1>-<t2>-<YYYY-MM-DD>-<market...>`;
the date is the match boundary and the only market-type-free stem. Dateless markets (elections,
`btc-updown-5m-<epoch>`) have no shared stem to over-merge, so they pass through 1:1 (never LESS
strict than the incumbent key there). Doubleheaders (same teams, same date, two games) would
merge — accepted, vanishingly rare and CONSERVATIVE (fewer independent events, not more).

Self-test:  ./superkey.py --self-test    (exercises every branch; exits non-zero on any failure)
"""

import re
import sys

_DATE = re.compile(r"(\d{4}-\d{2}-\d{2})")


def super_event(event_slug, slug):
    """Return the match-level cluster key. event_slug may be '' / None; slug is the fallback."""
    base = (event_slug or "").strip()
    if not base:
        base = (slug or "").strip()
    if not base:
        return None  # caller decides (e.g. fall back to condition_id) — matches incumbent behaviour
    m = _DATE.search(base)
    if not m:
        return base  # dateless market: 1:1, never coarser than the incumbent key
    return base[: m.end()]  # stem through the first date; drop the market-type suffix


# --- self-test -------------------------------------------------------------------------------
_CASES = [
    # (event_slug, slug, expected, why)
    ("fifwc-bel-sen-2026-07-01", "fifwc-bel-sen-2026-07-01-bel",
     "fifwc-bel-sen-2026-07-01", "game-winner already at match level"),
    ("fifwc-bel-sen-2026-07-01-exact-score", "fifwc-bel-sen-2026-07-01-exact-score-2-0",
     "fifwc-bel-sen-2026-07-01", "exact-score sub-market strips to match"),
    ("fifwc-bel-sen-2026-07-01-more-markets", "fifwc-bel-sen-2026-07-01-spread-home-1pt5",
     "fifwc-bel-sen-2026-07-01", "more-markets sub-market strips to match"),
    ("", "fifwc-bel-sen-2026-07-01-total-1pt5",
     "fifwc-bel-sen-2026-07-01", "empty event_slug falls back to slug then strips"),
    ("atp-jong-hijikat-2026-06-29", "atp-jong-hijikat-2026-06-29-set-handicap-away-1pt5",
     "atp-jong-hijikat-2026-06-29", "tennis moneyline already 1:1 (unchanged)"),
    ("mlb-laa-sea-2026-06-30", "mlb-laa-sea-2026-06-30-total-5pt5",
     "mlb-laa-sea-2026-06-30", "mlb total's event_slug already the match"),
    ("co-01-democratic-primary-winner", "will-diana-degette-be-the-nominee",
     "co-01-democratic-primary-winner", "dateless election market stays 1:1"),
    ("btc-updown-5m-1782784500", "btc-updown-5m-1782784500",
     "btc-updown-5m-1782784500", "dateless crypto up/down stays 1:1"),
    ("lol-t1-tl2-2026-07-01", "lol-t1-tl2-2026-07-01-game3",
     "lol-t1-tl2-2026-07-01", "esports maps of one series collapse to the series"),
    (None, "fifwc-civ-nor-2026-06-30-exact-score-3-0",
     "fifwc-civ-nor-2026-06-30", "None event_slug + slug fallback"),
]


def _self_test():
    ok = True
    # 1. every documented branch
    for evt, slug, want, why in _CASES:
        got = super_event(evt, slug)
        flag = "ok" if got == want else "FAIL"
        if got != want:
            ok = False
        print(f"  [{flag}] super_event({evt!r},{slug!r}) = {got!r}  ({why})")
    # 2. collapse property: the 3 Belgium-Senegal event_slugs + the empty-slug row → ONE key
    bel = {
        super_event("fifwc-bel-sen-2026-07-01", "x"),
        super_event("fifwc-bel-sen-2026-07-01-exact-score", "x"),
        super_event("fifwc-bel-sen-2026-07-01-more-markets", "x"),
        super_event("", "fifwc-bel-sen-2026-07-01-total-1pt5"),
    }
    if len(bel) != 1:
        ok = False
    print(f"  [{'ok' if len(bel)==1 else 'FAIL'}] 4 Belgium-Senegal slugs collapse to {len(bel)} key")
    # 3. NEVER coarser than incumbent on dateless keys (distinct dateless slugs stay distinct)
    a = super_event("co-01-democratic-primary-winner", "")
    b = super_event("colorado-governor-democratic-primary-winner", "")
    if a == b:
        ok = False
    print(f"  [{'ok' if a!=b else 'FAIL'}] two distinct dateless markets stay distinct")
    # 4. idempotence: super_event(super_event(x)) == super_event(x)
    idem = all(super_event(super_event(e, s), "") == super_event(e, s) for e, s, _, _ in _CASES)
    if not idem:
        ok = False
    print(f"  [{'ok' if idem else 'FAIL'}] idempotent")
    print("SELF-TEST", "PASS" if ok else "FAIL")
    return ok


if __name__ == "__main__":
    if "--self-test" in sys.argv:
        sys.exit(0 if _self_test() else 1)
    # filter mode: read "event_slug\tslug" lines, print the super key
    for line in sys.stdin:
        parts = line.rstrip("\n").split("\t")
        evt = parts[0] if len(parts) > 0 else ""
        slug = parts[1] if len(parts) > 1 else ""
        print(super_event(evt, slug))

File: common/src/superkey.rs
//! MATCH-LEVEL SUPER-KEY — the honest event-clustering key.
//!
//! A Rust mirror of `scripts/superkey.py::super_event` (the parity oracle). The
//! incumbent cluster key is `COALESCE(event_slug, condition_id)`, but `event_slug`
//! only groups the OUTCOME variants of ONE sub-market. A single real-world match
//! spans several `event_slug`s (game-winner, exact-score, more-markets, …), so
//! clustering on `event_slug` treats one match as up to ~4 "independent events"
//! and inflates N. The honest unit is the MATCH.
//!
//! Rule (documented, deterministic, NO fitting — identical to the Python):
//!   base = event_slug if non-empty else slug
//!   if base contains a YYYY-MM-DD date:
//!       super_key = base up-to-and-INCLUDING the first date
//!   else:
//!       super_key = base                        (dateless markets stay 1:1)
//!
//! Parity with `superkey.py --self-test` is unit-tested below.

/// Return the match-level cluster key. `event_slug` may be empty/None; `slug` is
/// the fallback. Returns `None` when both are empty (the caller then falls back
/// to `condition_id`, matching the incumbent behaviour).
pub fn super_event(event_slug: Option<&str>, slug: &str) -> Option<String> {
    let mut base = event_slug.unwrap_or("").trim();
    if base.is_empty() {
        base = slug.trim();
    }
    if base.is_empty() {
        return None;
    }
    match first_date_end(base) {
        // stem through (and including) the first YYYY-MM-DD; drop the market-type suffix
        Some(end) => Some(base[..end].to_string()),
        // dateless market: 1:1, never coarser than the incumbent key
        None => Some(base.to_string()),
    }
}

/// Byte offset just past the first `\d{4}-\d{2}-\d{2}` substring, or `None`.
/// Mirrors Python's `re.search(r"(\d{4}-\d{2}-\d{2})", base).end()`. ASCII slugs,
/// so byte offsets == char offsets.
fn first_date_end(s: &str) -> Option<usize> {
    let b = s.as_bytes();
    // pattern length is 10: DDDD-DD-DD
    if b.len() < 10 {
        return None;
    }
    for i in 0..=(b.len() - 10) {
        if b[i].is_ascii_digit()
            && b[i + 1].is_ascii_digit()
            && b[i + 2].is_ascii_digit()
            && b[i + 3].is_ascii_digit()
            && b[i + 4] == b'-'
            && b[i + 5].is_ascii_digit()
            && b[i + 6].is_ascii_digit()
            && b[i + 7] == b'-'
            && b[i + 8].is_ascii_digit()
            && b[i + 9].is_ascii_digit()
        {
            return Some(i + 10);
        }
    }
    None
}

#[cfg(test)]
mod tests {
    use super::*;

    // Parity cases copied verbatim from superkey.py `_CASES`:
    // (event_slug, slug, expected, why)
    #[test]
    fn parity_with_python_self_test() {
        let cases: &[(Option<&str>, &str, &str)] = &[
            (
                Some("fifwc-bel-sen-2026-07-01"),
                "fifwc-bel-sen-2026-07-01-bel",
                "fifwc-bel-sen-2026-07-01",
            ),
            (
                Some("fifwc-bel-sen-2026-07-01-exact-score"),
                "fifwc-bel-sen-2026-07-01-exact-score-2-0",
                "fifwc-bel-sen-2026-07-01",
            ),
            (
                Some("fifwc-bel-sen-2026-07-01-more-markets"),
                "fifwc-bel-sen-2026-07-01-spread-home-1pt5",
                "fifwc-bel-sen-2026-07-01",
            ),
            (
                Some(""),
                "fifwc-bel-sen-2026-07-01-total-1pt5",
                "fifwc-bel-sen-2026-07-01",
            ),
            (
                Some("atp-jong-hijikat-2026-06-29"),
                "atp-jong-hijikat-2026-06-29-set-handicap-away-1pt5",
                "atp-jong-hijikat-2026-06-29",
            ),
            (
                Some("mlb-laa-sea-2026-06-30"),
                "mlb-laa-sea-2026-06-30-total-5pt5",
                "mlb-laa-sea-2026-06-30",
            ),
            (
                Some("co-01-democratic-primary-winner"),
                "will-diana-degette-be-the-nominee",
                "co-01-democratic-primary-winner",
            ),
            (
                Some("btc-updown-5m-1782784500"),
                "btc-updown-5m-1782784500",
                "btc-updown-5m-1782784500",
            ),
            (
                Some("lol-t1-tl2-2026-07-01"),
                "lol-t1-tl2-2026-07-01-game3",
                "lol-t1-tl2-2026-07-01",
            ),
            (
                None,
                "fifwc-civ-nor-2026-06-30-exact-score-3-0",
                "fifwc-civ-nor-2026-06-30",
            ),
        ];
        for (evt, slug, want) in cases {
            assert_eq!(
                super_event(*evt, slug).as_deref(),
                Some(*want),
                "super_event({evt:?}, {slug:?})"
            );
        }
    }

    #[test]
    fn belgium_senegal_slugs_collapse_to_one_key() {
        let keys: std::collections::HashSet<_> = [
            super_event(Some("fifwc-bel-sen-2026-07-01"), "x"),
            super_event(Some("fifwc-bel-sen-2026-07-01-exact-score"), "x"),
            super_event(Some("fifwc-bel-sen-2026-07-01-more-markets"), "x"),
            super_event(Some(""), "fifwc-bel-sen-2026-07-01-total-1pt5"),
        ]
        .into_iter()
        .collect();
        assert_eq!(keys.len(), 1);
    }

    #[test]
    fn distinct_dateless_markets_stay_distinct() {
        let a = super_event(Some("co-01-democratic-primary-winner"), "");
        let b = super_event(Some("colorado-governor-democratic-primary-winner"), "");
        assert_ne!(a, b);
    }

    #[test]
    fn idempotent() {
        for (evt, slug) in [
            (Some("fifwc-bel-sen-2026-07-01-exact-score"), "x"),
            (Some("mlb-laa-sea-2026-06-30"), "y"),
            (Some("btc-updown-5m-1782784500"), "z"),
        ] {
            let once = super_event(evt, slug);
            let twice = super_event(once.as_deref(), "");
            assert_eq!(once, twice);
        }
    }

    #[test]
    fn both_empty_returns_none() {
        assert_eq!(super_event(None, ""), None);
        assert_eq!(super_event(Some("  "), "  "), None);
    }
}
